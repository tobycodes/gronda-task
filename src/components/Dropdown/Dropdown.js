import React, { useEffect, useState } from "react";

//I created a custom dropdown so we can reuse it for sorting, filtering and any other UI functionality that is similar to them
const Dropdown = ({ label, options, onChange }) => {
	const [isOpen, setIsOpen] = useState(false);
	const [selected, setSelected] = useState(options && options[0]); //Sets the first option in our options prop as the default sort/filtering option

	useEffect(() => {
		if (onChange) {
			//Any function passed as the onChange prop will be called with the currently selected option and can use it to request new data or modify UI appropriately.
			onChange(selected);
		}
	}, [selected, onChange]);

	const toggleDropdown = () => setIsOpen(!isOpen); //This function controls the opening/closing of the dropdown

	const handleSelected = (option) => {
		setSelected(option);
		setIsOpen(false);
	};

	return (
		<div className="dropdown">
			<span className="dropdown-label">{label}</span>
			<input
				className="dropdown-button"
				type="button"
				value={selected.title}
				onClick={toggleDropdown}
			/>

			<ul id="dropdown-list" className={`dropdown-list${isOpen ? " show" : ""}`}>
				{options.map((option) => (
					<li
						className="dropdown-item"
						key={option.id}
						onClick={() => handleSelected(option)} //This function remembers the particular option passed via closures, and we use that to set selected item on click.
					>
						{option.title}
					</li>
				))}
			</ul>
		</div>
	);
};

export default Dropdown;
