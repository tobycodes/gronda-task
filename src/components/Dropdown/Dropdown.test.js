import React from "react";
import { shallow } from "enzyme";

import Dropdown from "./Dropdown";

const props = {
	label: "Filter by",
	options: [
		{
			id: 1,
			title: "This month",
			time_unit: "MONTH",
			time_unit_count: 0,
		},
	],
	onChange: jest.fn(() => {}),
};

describe("<Dropdown />", () => {
	it("expect to render a Dropdown component", () => {
		expect(shallow(<Dropdown {...props} />)).toMatchSnapshot();
	});

	it("expect to render a button and a list", () => {
		const wrapper = shallow(<Dropdown {...props} />);

		expect(wrapper.find('input[type="button"]').length).toEqual(1);
		expect(wrapper.find("ul>li").length).toBeGreaterThanOrEqual(1);
	});

	it("correctly toggle the dropdown open or close", () => {
		const wrapper = shallow(<Dropdown {...props} />);

		//Simulate first click on button to open the dropdown.
		wrapper.find('input[type="button"]').simulate("click");
		expect(wrapper.find("#dropdown-list").props()).toHaveProperty(
			"className",
			"dropdown-list show"
		);

		//Simulate second click to close the dropdown.
		wrapper.find('input[type="button"]').simulate("click");
		expect(wrapper.find("#dropdown-list").props()).toHaveProperty(
			"className",
			"dropdown-list"
		);
	});
});
