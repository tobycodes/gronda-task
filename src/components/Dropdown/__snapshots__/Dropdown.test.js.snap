// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> expect to render a Dropdown component 1`] = `ShallowWrapper {}`;
